portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java

```diff
@@ -1425,6 +1432,71 @@ protected void validateFolderName(String folderName)
 		}
 	}
 
+	private void _propagateRolesPermissions(DLFolder dlFolder)
+		throws PortalException {
+
+		long companyId = dlFolder.getCompanyId();
+
+		long parentFolderId = dlFolder.getParentFolderId();
+
+		String parentClassName = DLFolderConstants.getClassName();
+
+		if (parentFolderId == DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
+			int count =
+				_resourcePermissionLocalService.getResourcePermissionsCount(
+					companyId, DLConstants.SERVICE_NAME,
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(dlFolder.getGroupId()));
+
+			if (count == 0) {
+				return;
+			}
+
+			parentClassName = DLConstants.RESOURCE_NAME;
+			parentFolderId = dlFolder.getGroupId();
+		}
+
+		// TODO Update the check permission in here
+
+		PermissionPropagationEntry permissionPropagationEntry =
+			_permissionPropagationEntryLocalService.
+				fetchPermissionPropagationEntry(
+					dlFolder.getCompanyId(), dlFolder.getGroupId(),
+					DLFolderConstants.getClassName(), parentFolderId);
+
+		if (permissionPropagationEntry == null) {
+			return;
+		}
+
+		Set<String> dlFolderActionIds = SetUtil.fromCollection(
+			ResourceActionsUtil.getModelResourceActions(
+				DLFolderConstants.getClassName()));
+
+		Map<Long, Set<String>> roleIdsToActionIds =
+			_resourcePermissionLocalService.
+				getAvailableResourcePermissionActionIds(
+					companyId, parentClassName,
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(parentFolderId), dlFolderActionIds);
```

Comment:
In the Home folder, I think the subfolder should inherit `ADD_FOLDER` permissions and map it to `ADD_SUBFOLDER`. Please help double-check.